### DISCUSSION TITLE
[ARCHIVED] Entities в методологии и их связность

### DISCUSSION THREAD

**Discussion by Lehakos** [feature-sliced/documentation#49] [Archive]
Я понимаю, что entities, так же как и фичи, не должны импортировать друг друга. Но вот как этого избегать? Допустим, я делаю игру типо "Мафии". У меня есть сущность Player и есть сущность Role. Понятно, что в сущности Player есть поле role. А также может быть компонент PlayerCard, в котором помимо информации о самом игроке должна выводиться информация о его роли

**Comment by sergeysova** [comment]
Кажется, что не стоит пытаться переложить feature-sliced на игровые механики и сущности.
Лучше рассматривать на практических примерах фронтенд-проектов.

**Reply by azinit** [reply]
@sergeysova думаю вопрос хороший все же)

Т.к. вопросы
- "у меня фича импортит другую фичу"
- "А здесь у меня один энтити обращается напрямую к другому"
- и т.д.

Очень сильно коррелируют на самом деле с нашей методологией)

**Reply by azinit** [reply]
Даже мы со своим [feature-driven](https://www.youtube.com/watch?v=BWAeYuWFHhs) задавались вопросом - "А как этого избежать?"

И чаще всего подходящий ответ - "Вынести связи на уровень выше"

> См также ответ [Oleg Isonen](https://github.com/kof/feature-driven-architecture/issues/11#issuecomment-727911392)

- Для нескольких связных фичей - это вынести связь на уровень страницы/процесса
- Для нескольких связных энтити - соответственно, на уровень фичей/страницы и т.д.

**Reply by azinit** [reply]
@sergeysova тут на самом деле другой вопрос - "А что если у нас вложенные entities?"

Опять таки аналог вопроса "А что если у нас фича в другой фиче?"

**Reply by sergeysova** [reply]
Опять же, лучше практический пример смотреть

**Reply by azinit** [reply]
> Кажется, что не стоит пытаться переложить feature-sliced на игровые механики и сущности.
> Лучше рассматривать на практических примерах фронтенд-проектов.

Потому что в том числе можно переложить этот пример с игры на реальный

> Есть пользователь (User) и его карточка
> И есть его "Репутация/Роль/{YOUR_ENTITY_NAME}" - который надо тоже отобразить, но тем не менее он внутри пользователя

**Reply by Lehakos** [reply]
> Кажется, что не стоит пытаться переложить feature-sliced на игровые механики и сущности.
> Лучше рассматривать на практических примерах фронтенд-проектов.

А разве нельзя писать игру на реакте? Я, например, делаю для себя сейчас. Такую как Мафия почему нет?

**Reply by Lehakos** [reply]
> Даже мы со своим [feature-driven](https://www.youtube.com/watch?v=BWAeYuWFHhs) задавались вопросом - "А как этого избежать?"
> 
> И чаще всего подходящий ответ - "Вынести связи на уровень выше"

Я думаю интерфейсы можно выносить в отдельную директорию на одном уровне. Допустим entities/types или interfaces. Тогда можно будет хотя бы интерфейсы импортировать на одном уровне, будет работать dependency inversion principle. А уже компоненты компонавать можно на уровень выше

**Reply by ilyaagarkov** [reply]
> А разве нельзя писать игру на реакте? Я, например, делаю для себя сейчас. Такую как Мафия почему нет?

на react писать можно конечно. React это лишь инструмент для отображения. Тут речь именно про feature-sliced, и что возможно не стоит пытаться натянуть эту методологию на игровые механики.

**Reply by azinit** [reply]
@ilyaagarkov , @sergeysova  **`feature-sliced` - про разрезание большого и сложного проекта, как по мне**

*Понятно, что с игровой разработкой добавляется еще контекст понятий, например Сцены, Модели (3D), Шейдеры, Материалы, Probes и т.д.*

Но feature-sliced в том числе задается вопросами 
- "А как разрезать?" 
- "А что если нам нужно заимпортить фичу/entity/... внутри другой?
- и т.д.


И потому, мы, как приверженцы методологии, должны задумываться и о таких вещах!

> @ilyaagarkov тем более ты сам чуть ниже про это написал (https://github.com/feature-sliced/wiki/discussions/49#discussioncomment-477672)

**Comment by ilyaagarkov** [comment]
> > @sergeysova тут на самом деле другой вопрос - "А что если у нас вложенные entities?"
> 
> Опять же, лучше практический пример смотреть

Думаю, что показательный пример - Пост с Комментариями.  
Куда складывать компонент для отображения? Сразу учтем случай что комментарии могут быть не только у поста.

Кажется, что все таки entities должны импортировать другие entities. Ведь у нас всегда есть связи между сущностями, 

> И чаще всего подходящий ответ - "Вынести связи на уровень выше"

если брать те уровни которые у нас есть, то я не вижу где бы это могло быть.

**Reply by ecrofeg** [reply]
Использую FSD не на первом, и даже не на втором большом проекте (e-commerce), и вопрос о том "можно ли импортировать сущности друг в друга" встает всегда, потому что в реальном мире все сущности взаимосвязаны, какую бизнес-область ни бери :) 

@azinit тред старый, но может подскажешь, какая сейчас основная рекомендация на этот счет? может стоит про это как-то упомянуть в доке (вроде бы про это ничего нет в разделе https://feature-sliced.design/ru/docs/reference/units/layers/entities)?

Вероятно даже эта страница лучше подходит https://feature-sliced.design/ru/docs/reference/isolation/decouple-entities

**Reply by azinit** [reply]
@feature-sliced/core

**Reply by aau8** [reply]
Ну, а как же shared? Одним из решений данной я проблемы, я вижу перенос сущностей (моделей) в каталог shared, где каждая сущность сможет импортировать любую другую, но тогда в entities останется только api и ui, а самой сущности не будет, что не совсем соответствует наименованию каталога...

**Comment by vad1ym** [comment]
Может связывать все в виджете? Я пока только вникаю в методологию, но виджет вроде бы подходит для данной цели:

```
/// widgets/PlayerCardWidget
import { PlayerCard, userModel } from 'entities/user';
import { RoleBadge } from 'entities/role';

<PlayerCard role={<RoleBadge role="userModel.role" /> } />
```

Ну а если роль это комплексный объект, то реализовать возможность в роль-моедел получить роль по имени:

```
/// widgets/PlayerCardWidget
import { PlayerCard, userModel } from 'entities/user';
import { RoleBadge, roleModel } from 'entities/role';

<PlayerCard role={<RoleBadge role="roleModel.getRoleByName(userModel.role)" /> } />
```
